core/src/config/substrate.rs

```diff
@@ -225,12 +227,69 @@ where
 {
     // At the time of writing, Smoldot gives back block numbers in numeric rather
     // than hex format. So let's support deserializing from both here:
-    use crate::backend::legacy::rpc_methods::NumberOrHex;
     let number_or_hex = NumberOrHex::deserialize(d)?;
     let u256 = number_or_hex.into_u256();
     TryFrom::try_from(u256).map_err(|_| serde::de::Error::custom("Try from failed"))
 }
 
+/// A number type that can be serialized both as a number or a string that encodes a number in a
+/// string.
+///
+/// We allow two representations of the block number as input. Either we deserialize to the type
+/// that is specified in the block type or we attempt to parse given hex value.
+///
+/// The primary motivation for having this type is to avoid overflows when using big integers in
+/// JavaScript (which we consider as an important RPC API consumer).
+#[derive(Copy, Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
+#[serde(untagged)]
```

Comment:
This type is only needed for the RPC stuff and I wonder whether we could move to the RPC stuff instead? 

IIRC, the RPC stuff doesn't compile for no-std and that's the reason why to separate it

Reply:
Yeah, but the problem I have is that the `SubstrateConfig` should be defined in `subxt-core` and this has the associated type `Header = SubstrateHeader<32, _>` which has a field, `number: u32,` that is tagged with 
```
[serde(
        serialize_with = "serialize_number",
        deserialize_with = "deserialize_number"
)]
```
where `deserialize_number` uses `NumberOrHex`. I am not sure if I can easily move all of these Serialize implmentations to the RPC stuff, especially because the `SubstrateHeader` is now in a different crate.

Reply:
I'm ok with it being here; it _is_ an RPC type but it's also just a specific concrete type we decide to use in our `SubstrateConfig` impl, so it should live next to it I reckon :)